### PR TITLE
multithread / multiprocess safe dataset reads

### PR DESCRIPTION
**Pull request by rlizzo**
## Motivation and Context
Why is this change required? What problem does it solve?:

This is the initial work on making hangar reads thread/process safe. The `batch_get` function is the preffered method for user interaction. 

If it fixes an open issue, please link to the issue here:

Addresses some of the requirements to enable: #13 

## Description
Describe your changes in detail:

`DatasetDataReader` class no longer contains references to lmdb environments. This change was made as it cannot be serialized and sent across processes. 

## Types of changes
What types of changes does your code introduce? Put an `x` in all the boxes that apply:
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

Is this PR ready for review, or a work in progress?
- [ ] Ready for review
- [x] Work in progress

## How Has This Been Tested?
Put an `x` in the boxes that apply:
- [ ] Current tests cover modifications made
- [ ] New tests have been added to the test suite
- [ ] Modifications were made to existing tests to support these changes
- [x] Tests may be needed, but they are not included when the PR was proposed
- [ ] I don't know. Help!

## Checklist:
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **[CONTRIBUTING](../CONTRIBUTING.rst)** document.
- [x] I have signed (or will sign when prompted) the tensorwork CLA.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: src/hangar/dataset.py [hunk 1]
```diff
@@ -239,66 +228,105 @@ def iswriteable(self):
...
-        '''Retrieve a dataset data sample with the provided sample name
+    def get(self, name: str) -> np.ndarray:
+        '''Retrieve a sample in the dataset with a specific name.
+
+        The method is thread/process safe IF used in a read only checkout. Use
+        this if the calling application wants to manually manage multiprocess
+        logic for data retrieval. Otherwise, hangar includes the ``batch_get``
+        method to retrieve multiple data samples simultaneously. This method
+        uses multiprocess pool of workers (managed by hangar) to drastically
+        increase access speed and simplifly application developer workflows.
```
Comment (hhsecond) [inline src/hangar/dataset.py]: simplify?

File: src/hangar/dataset.py [hunk 2]
```diff
@@ -239,66 +228,105 @@ def iswriteable(self):
...
             yield (name, self.get(name))
 
-    def get(self, name):
-        '''Retrieve a dataset data sample with the provided sample name
+    def get(self, name: str) -> np.ndarray:
+        '''Retrieve a sample in the dataset with a specific name.
+
+        The method is thread/process safe IF used in a read only checkout. Use
+        this if the calling application wants to manually manage multiprocess
+        logic for data retrieval. Otherwise, hangar includes the ``batch_get``
```
Comment (hhsecond) [inline src/hangar/dataset.py]: `get_batch`?

**Comment by rlizzo** [comment]
Merging now!

**Comment by rlizzo** [comment]
Thanks for the review @hhsecond!

**Pull request merged**